<a href="https://colab.research.google.com/github/RepnaiaE/DataScience_Emotion-recognition-methods/blob/main/Lab4_Classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Классификация. Метод опорных векторов (SVM), ядерный SVM. Байесовский классификатор. Классификатор на основе К ближайших соседей (KNN). Деревья решений. Упрощение деревьев. Случайный лес. Оценка эффективности и сравнение моделей. /Classification. Support vector machine (SVM), kernel SVM. Bayesian classifier. Classifier based on K nearest neighbors (KNN). decision trees. Simplifying trees. Random forest. Evaluation of efficiency and comparison of models.

# Загрузка данных/Loading data

Загружаем данные из ячеек в массивы признаков для каждого набора из предыдущей работы

In [ ]:
import pandas as pd
import numpy as np

ling_features = pd.read_csv(r'C:\Users\voloshina\Documents\Notebooks\Распознавание эмоций\Распознавание эмоций\ru_tweet_corp_ling_features.csv').sample(frac=1)
ling_features_vec = list(ling_features['features'].apply(lambda y: [float(x) for x in y[1:len(y)-1].split(',')]))

In [ ]:
morph_features = pd.read_csv(r'C:\Users\voloshina\Documents\Notebooks\Распознавание эмоций\Распознавание эмоций\ru_tweet_corp_morph_features.csv').sample(frac=1)
morph_features_vec = list(morph_features['coded_pos'].apply(lambda y: [float(x) for x in y[1:len(y)-1].split(',')]))

In [ ]:
tfidf_features = pd.read_csv(r'C:\Users\voloshina\Documents\Notebooks\Распознавание эмоций\Распознавание эмоций\ru_tweet_corp_tfidf.csv').sample(frac=1)
tfidf_features_vec = list(tfidf_features['tfidf'].apply(lambda y: [float(x) for x in y[1:len(y)-1].split(',')]))

In [ ]:
import pandas as pd


concated_features = pd.read_csv(r'C:\Users\voloshina\Documents\Notebooks\Распознавание эмоций\Распознавание эмоций\ru_tweet_corp_tfidf.csv').sample(frac=1)

features_tfidf = list(concated_features['tfidf'].apply(lambda y: [float(x) for x in y[1:len(y)-1].split(',')]))
morph_features_coded_pos = list(concated_features['coded_pos'].apply(lambda y: [float(x) for x in y[1:len(y)-1].split(',')]))
ling_features_features = list(concated_features['features'].apply(lambda y: [float(x) for x in y[1:len(y)-1].split(',')]))

concated_features_vec = []
for tf, morph, ling in zip(features_tfidf, morph_features_coded_pos, ling_features_features):
    concated_features_vec.append(tf + morph + ling)

In [ ]:
len(concated_features_vec[0])

1023

# Классификаторы/Classifiers

In [ ]:
from sklearn import svm
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.naive_bayes import MultinomialNB
from sklearn.neighbors import KNeighborsClassifier

from sklearn.model_selection import GridSearchCV

from sklearn.preprocessing import Normalizer
from sklearn.preprocessing import MinMaxScaler

from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import precision_score
from sklearn.metrics import f1_score
from sklearn.metrics import make_scorer


def calc_clfs(features, y, test_size=200):
    X = features[:len(features) - test_size]
    y = list(y)[:len(y) - test_size]

    for_pred_x = features[len(features) - test_size:len(features)]
    expected_y = list(y)[len(y) - test_size:len(y)]

    svm_clf = svm.SVC()
    svm_clf.fit(X, y)
    y_pred = svm_clf.predict(for_pred_x)
#     print(f'svm prediction {y_pred}, expected {expected_y}')
    print('metrics for svm:')
    print(f'accuracy = {accuracy_score(expected_y, y_pred)}')
    print(f'recall = {recall_score(expected_y, y_pred)}')
    print(f'precision = {precision_score(expected_y, y_pred)}')
    print(f'f1 = {f1_score(expected_y, y_pred)}')

    print()

    tree_clf = DecisionTreeClassifier(criterion='entropy', max_depth=10, random_state=0)
    tree_clf.fit(X, y)
    y_pred = tree_clf.predict(for_pred_x)
#     print(f'tree prediction {y_pred}, expected {expected_y}')
    print('metrics for tree:')
    print(f'accuracy = {accuracy_score(expected_y, y_pred)}')
    print(f'recall = {recall_score(expected_y, y_pred)}')
    print(f'precision = {precision_score(expected_y, y_pred)}')
    print(f'f1 = {f1_score(expected_y, y_pred)}')

    print()

    forest_clf = RandomForestClassifier(max_depth=5, random_state=0)
    forest_clf.fit(X, y)
    y_pred = forest_clf.predict(for_pred_x)
#     print(f'forest prediction {y_pred}, expected {expected_y}')
    print('metrics for forest:')
    print(f'accuracy = {accuracy_score(expected_y, y_pred)}')
    print(f'recall = {recall_score(expected_y, y_pred)}')
    print(f'precision = {precision_score(expected_y, y_pred)}')
    print(f'f1 = {f1_score(expected_y, y_pred)}')

    print()

#     transformer = Normalizer(norm='max').fit(X)
#     XX =  transformer.transform(X)
#     XX = [list(map(abs, x)) for x in X]
    scaler = MinMaxScaler()
    XX = scaler.fit_transform(X)

    bayes_clf = MultinomialNB()
    bayes_clf.fit(XX, y)
    y_pred = bayes_clf.predict(for_pred_x)
#     print(f'bayes prediction {y_pred}, expected {expected_y}')
    print('metrics for bayes:')
    print(f'accuracy = {accuracy_score(expected_y, y_pred)}')
    print(f'recall = {recall_score(expected_y, y_pred)}')
    print(f'precision = {precision_score(expected_y, y_pred)}')
    print(f'f1 = {f1_score(expected_y, y_pred)}')

    print()

    knn_clf = KNeighborsClassifier(n_neighbors=5)
    knn_clf.fit(X, y)
    y_pred = knn_clf.predict(for_pred_x)
#     print(f'KNN prediction {y_pred}, expected {expected_y}')
    print('metrics for knn:')
    print(f'accuracy = {accuracy_score(expected_y, y_pred)}')
    print(f'recall = {recall_score(expected_y, y_pred)}')
    print(f'precision = {precision_score(expected_y, y_pred)}')
    print(f'f1 = {f1_score(expected_y, y_pred)}')

    print()

    f1 = make_scorer(f1_score , average='macro')

    parameters = {'kernel':('linear', 'rbf'),
                  'C':[i for i in range(1, 10, 2)]}
    clf_svm = GridSearchCV(svm_clf, parameters, scoring=f1)
    clf_svm.fit(X, y)
    print('for svm: ')
    print(f'best score {clf_svm.best_score_}')
    print(f'best params {clf_svm.best_params_}')

    print()

    parameters = {'criterion':('gini', 'entropy'),
                  'max_depth':[i for i in range(3, 10, 2)]}
    clf_tree = GridSearchCV(tree_clf, parameters, scoring=f1)
    clf_tree.fit(X, y)
    print('for tree: ')
    print(f'best score {clf_tree.best_score_}')
    print(f'best params {clf_tree.best_params_}')

    print()

    parameters = {'criterion':('gini', 'entropy'),
                  'max_depth':[i for i in range(2, 10, 2)],
                  'n_estimators':[i for i in range(10, 100, 10)]}
    clf_forest = GridSearchCV(forest_clf, parameters, scoring=f1)
    clf_forest.fit(X, y)
    print('for forest: ')
    print(f'best score {clf_forest.best_score_}')
    print(f'best params {clf_forest.best_params_}')

    print()

    parameters = {'alpha':[0.0, 0.2, 0.5, 0.7, 1.0]}
    clf_bayes = GridSearchCV(bayes_clf, parameters, scoring=f1)
    clf_bayes.fit(XX, y)
    print('for bayes: ')
    print(f'best score {clf_bayes.best_score_}')
    print(f'best params {clf_bayes.best_params_}')

    print()

    parameters = {'weights':('uniform', 'distance'),
                  'algorithm':('auto', 'kd_tree', 'brute'),
                  'n_neighbors':[i for i in range(1, 10, 1)]}
    clf_knn = GridSearchCV(knn_clf, parameters, scoring=f1)
    clf_knn.fit(X, y)
    print('for knn: ')
    print(f'best score {clf_knn.best_score_}')
    print(f'best params {clf_knn.best_params_}')

Лингвистические признаки/Linguistic features

In [ ]:
calc_clfs(ling_features_vec, ling_features['ttype'].values)

metrics for svm:
accuracy = 0.5
recall = 0.7387387387387387
precision = 0.5359477124183006
f1 = 0.6212121212121212

metrics for tree:
accuracy = 0.525
recall = 0.6756756756756757
precision = 0.5597014925373134
f1 = 0.6122448979591837

metrics for forest:
accuracy = 0.515
recall = 0.7027027027027027
precision = 0.5492957746478874
f1 = 0.6166007905138341

metrics for bayes:
accuracy = 0.5
recall = 0.6846846846846847
precision = 0.5390070921985816
f1 = 0.6031746031746031

metrics for knn:
accuracy = 0.44
recall = 0.24324324324324326
precision = 0.4909090909090909
f1 = 0.3253012048192771

for svm: 
best score 0.5359167739156953
best params {'C': 3, 'kernel': 'linear'}

for tree: 
best score 0.4891658372985336
best params {'criterion': 'gini', 'max_depth': 7}

for forest: 
best score 0.5264853838335243
best params {'criterion': 'gini', 'max_depth': 4, 'n_estimators': 10}

for bayes: 
best score 0.5141906682695093
best params {'alpha': 0.0}



A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too small will result in numeric errors, '
A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too small will result in numeric errors, '
A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too small will result in numeric errors, '
A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too small will result in numeric errors, '
A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too

for knn: 
best score 0.5385413182880072
best params {'algorithm': 'brute', 'n_neighbors': 9, 'weights': 'uniform'}


Морфологические признаки/Morphological features

In [ ]:
calc_clfs(morph_features_vec, morph_features['ttype'].values)

metrics for svm:
accuracy = 0.515
recall = 0.5360824742268041
precision = 0.5
f1 = 0.5174129353233832

metrics for tree:
accuracy = 0.58
recall = 0.5670103092783505
precision = 0.5670103092783505
f1 = 0.5670103092783505

metrics for forest:
accuracy = 0.535
recall = 0.5360824742268041
precision = 0.52
f1 = 0.5279187817258884

metrics for bayes:
accuracy = 0.525
recall = 0.6288659793814433
precision = 0.5083333333333333
f1 = 0.5622119815668203

metrics for knn:
accuracy = 0.53
recall = 0.5154639175257731
precision = 0.5154639175257731
f1 = 0.5154639175257731

for svm: 
best score 0.5832741384962992
best params {'C': 7, 'kernel': 'rbf'}

for tree: 
best score 0.5434858127590749
best params {'criterion': 'entropy', 'max_depth': 9}

for forest: 
best score 0.5909152487440459
best params {'criterion': 'entropy', 'max_depth': 4, 'n_estimators': 90}

for bayes: 
best score 0.5776812380697776
best params {'alpha': 0.2}



A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too small will result in numeric errors, '
A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too small will result in numeric errors, '
A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too small will result in numeric errors, '
A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too small will result in numeric errors, '
A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too

for knn: 
best score 0.5570535248805879
best params {'algorithm': 'auto', 'n_neighbors': 1, 'weights': 'uniform'}


TF-IDF

In [ ]:
calc_clfs(tfidf_features_vec, tfidf_features['ttype'].values)

metrics for svm:
accuracy = 0.56
recall = 0.5894736842105263
precision = 0.5333333333333333
f1 = 0.5599999999999999

metrics for tree:
accuracy = 0.52
recall = 0.17894736842105263
precision = 0.4857142857142857
f1 = 0.26153846153846155

metrics for forest:
accuracy = 0.515
recall = 0.29473684210526313
precision = 0.4827586206896552
f1 = 0.3660130718954248

metrics for bayes:
accuracy = 0.55
recall = 0.6
precision = 0.5229357798165137
f1 = 0.5588235294117647

metrics for knn:
accuracy = 0.48
recall = 0.9578947368421052
precision = 0.47643979057591623
f1 = 0.6363636363636364

for svm: 
best score 0.6632875355153708
best params {'C': 1, 'kernel': 'rbf'}

for tree: 
best score 0.5126206043571685
best params {'criterion': 'gini', 'max_depth': 9}

for forest: 
best score 0.6089883593798066
best params {'criterion': 'gini', 'max_depth': 8, 'n_estimators': 90}



A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too small will result in numeric errors, '
A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too small will result in numeric errors, '
A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too small will result in numeric errors, '
A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too small will result in numeric errors, '
A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too

for bayes: 
best score 0.6498811250200893
best params {'alpha': 1.0}

for knn: 
best score 0.5145948493346595
best params {'algorithm': 'kd_tree', 'n_neighbors': 2, 'weights': 'distance'}


Concat 3 features

In [ ]:
calc_clfs(concated_features_vec, concated_features['ttype'].values)

metrics for svm:
accuracy = 0.5
recall = 0.46
precision = 0.5
f1 = 0.4791666666666667

metrics for tree:
accuracy = 0.53
recall = 0.62
precision = 0.5254237288135594
f1 = 0.5688073394495413

metrics for forest:
accuracy = 0.53
recall = 0.45
precision = 0.5357142857142857
f1 = 0.4891304347826087

metrics for bayes:
accuracy = 0.495
recall = 0.47
precision = 0.49473684210526314
f1 = 0.482051282051282

metrics for knn:
accuracy = 0.45
recall = 0.59
precision = 0.4609375
f1 = 0.5175438596491228

for svm: 
best score 0.663996620417442
best params {'C': 3, 'kernel': 'rbf'}

for tree: 
best score 0.5896492350940675
best params {'criterion': 'entropy', 'max_depth': 9}

for forest: 
best score 0.6503649680707259
best params {'criterion': 'entropy', 'max_depth': 8, 'n_estimators': 50}



A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too small will result in numeric errors, '
A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too small will result in numeric errors, '
A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too small will result in numeric errors, '
A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too small will result in numeric errors, '
A:\Anaconda\lib\site-packages\sklearn\naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  warnings.warn('alpha too

for bayes: 
best score 0.6531511310683333
best params {'alpha': 1.0}

for knn: 
best score 0.6132040584231034
best params {'algorithm': 'auto', 'n_neighbors': 8, 'weights': 'uniform'}


# Выводы/Conclusion

По результатам экспериментов лучшими подходами для классификации тональности по каждому набору признаков стали:

| Features | Model| Params | F1 score |
|----------|------|--------|----------|
| Лингвистические | SVM | без доп.настройки | 0.62 |
| Морфологические | Random forest | 'criterion': 'entropy', 'max_depth': 4, 'n_estimators': 90 | 0.59 |
| TF-IDF | SVM | 'C': 1, 'kernel': 'rbf' | 0.66 |
| Объединение признаков | SVM | 'C': 3, 'kernel': 'rbf' | 0.66 |

Как итог, хотя многие модели оказываются сопоставимы по качеству распознавания, SVM показывает чуть лучшие результаты.

Среди способов выделения признаков выделяется TF-IDF и объединение всех 3 векторов признаков. Хотя по результатам обучения кажется, что основной вклад в объединенном векторе внесли именно TF-IDF признаки и остались на том же уровне. Добавление новых признаков качество классификации не улучшило.

As a result, although many models are comparable in terms of recognition quality, SVM shows slightly better results.

Among the methods of feature extraction, TF-IDF and the union of all 3 feature vectors stand out. Although, according to the training results, it seems that the TF-IDF features made the main contribution to the combined vector and remained at the same level. The addition of new features did not improve the classification quality.